File: python/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1]
        
        for i in range(len(nums)-2, -1, -1):
            res.append(res[-1]*nums[i+1])
        
        res.reverse()
        
        curr = 1
        for i in range(len(nums)):
            res[i] *= curr
            curr *= nums[i]
        
        return res

File: go/product_of_array_except_self.go
package leetcode

func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))
	res[len(res)-1] = 1

	for i := len(nums) - 2; i >= 0; i-- {
		res[i] = res[i+1] * nums[i+1]
	}

	curr := 1
	for i := 0; i < len(nums); i++ {
		res[i] *= curr
		curr *= nums[i]
	}

	return res
}
